### PR TITLE
cEP-0035: Extending support for newer versions of Python

### LINKED ISSUE #191: cEP-0035: Extending support for newer versions of Python
This cEP describes the upgrades and steps required to enable coala to support newer versions of Python. This project has been taken up as a [GSoC](https://summerofcode.withgoogle.com/organizations/6490386047959040/#4634486723051520) project.

### PR DESCRIPTION
**Pull request by anshalshukla**
This proposal describes the approach that will be followed
to enable coala work with newer version of Python.

closes #191

### CODE REVIEW

File: cEP-0035.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Extending support for newer versions of Python
+
+| Metadata |                                       |
+| -------- | ------------------------------------- |
+| cEP      | 0035                                  |
+| Version  | 1.0                                   |
+| Title    | Extending support for newer version of Python  |
+| Authors  | Anshal Shukla <mailto:shukla.anshal85@gmail.com> |
```
Comment (abhishalya) [inline cEP-0035.md]: Could improve the formatting here a bit.

File: cEP-0035.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+| Status   | Proposed                              |
+| Type     | Process                               |
+
+## Abstract
+
+This cEP describes the upgrades and steps required to enable coala to support newer versions of Python. This project has been taken up as a [GSoC](https://summerofcode.withgoogle.com/organizations/6490386047959040/#4634486723051520) project.
```
Comment (abhishalya) [inline cEP-0035.md]: Mention the GSoC year, also this line is too long, please use a decent max line length.

File: cEP-0035.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+coala, as of now, supports python up to version 3.6, which is quite old. Lack of support of newer versions of python has been a significant hurdle in the development of coala. Version 3.6 was released back in 2016, and python has made a considerable amount of progress since then. coala, on the other hand, has largely not been able to benefit from these upgrades because of the strict version limitations. Upgrading coala to support newer versions will accelerate development of coala and will make it future proof.
+
+## Objectives
+
+The primary goal of the project is to enable coala to support newer versions of Python but this will be attained in multiple smaller steps. The goal is not just to enable support for newer versions of Python but also upgrade the coala to support newer versions of dependencies. Upgrades will be performed over multiple repositories as the dependencies across the project are closely interlinked. Changes will be performed in coala, bears, dependency_mangement, docker, mobans and coala-utils repository. 
+
+## Proposed Approach
+
+### 1. Keeping the CI/CD pipeline green
```
Comment (abhishalya) [inline cEP-0035.md]: Let's not use numbers in headings.

**Review by abhishalya — CHANGES_REQUESTED** [review]
This looks very vague in terms of the changes being made as a part of the project. It is important to pinpoint the changes as much as possible. A lot of which is already available in your project.

**Comment by abhishalya** [comment]
Please no more commits like `aa`, `aaa`. It's an extremely bad practice to write such commits.

**Comment by anshalshukla** [comment]
I had to change the commit message to "Add support for newer versions of Python" because having "Extending" makes the commit headlog imperative. coala throws an error with that. Let me know if I should change that project-wide.

File: cEP-0035.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
+# Extending support for newer versions of Python
+
+| Metadata |                                       |
+| -------- | ------------------------------------- |
+| cEP      | 0035                                  |
+| Version  | 1.0                                   |
+| Title    | Extending support for newer version of Python    |
+| Authors  | Anshal Shukla <mailto:shukla.anshal85@gmail.com> |
```
Comment (abhishalya) [inline cEP-0035.md:8]: You can improve formatting a bit here.

File: cEP-0035.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+| Status   | Proposed                              |
+| Type     | Process                               |
+
+## Abstract
+
+This cEP describes the upgrades and steps required to enable coala to support newer versions of Python. This project has been taken up as a [GSoC'21](https://summerofcode.withgoogle.com/organizations/6490386047959040/#4634486723051520) project.
```
Comment (abhishalya) [inline cEP-0035.md]: Group all links at the bottom and use them here.

File: cEP-0035.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+## Introduction
+
+coala, as of now, supports python up to version 3.6, which is quite old. Lack of support of newer versions of python has been a significant hurdle in the development of coala. Version 3.6 was released back in 2016, and python has made a considerable amount of progress since then. coala, on the other hand, has largely not been able to benefit from these upgrades because of the strict version limitations. Upgrading coala to support newer versions will accelerate development of coala and will make it future proof.
```
Comment (abhishalya) [inline cEP-0035.md]: Respect line length please.

File: cEP-0035.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+## Objectives
+
+TThe project's primary goal is to enable coala to support newer versions of Python, but this will be attained in multiple smaller steps. The goal is not just to facilitate support for newer versions of Python but also to upgrade the coala to support more recent dependencies. Upgrades will be performed over multiple repositories as the dependencies across the project are closely interlinked. Will perform changes in coala, bears, dependency_mangement, docker, mobans and coala-utils repository.
```
Comment (abhishalya) [inline cEP-0035.md]: coala-bears.

File: cEP-0035.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+## Objectives
+
+TThe project's primary goal is to enable coala to support newer versions of Python, but this will be attained in multiple smaller steps. The goal is not just to facilitate support for newer versions of Python but also to upgrade the coala to support more recent dependencies. Upgrades will be performed over multiple repositories as the dependencies across the project are closely interlinked. Will perform changes in coala, bears, dependency_mangement, docker, mobans and coala-utils repository. 
+
+## Proposed Approach
+
+### Keeping the CI/CD pipeline green
+
+Throughout the project, it needs to be ensured that the CI/CD pipeline remains unbroken as it is crucial for running the test suite, which is necessary for PR to get merged. This can be ensured by making commits keeping into consideration the pipeline. And make changes over multiple repositories before merging, in case required to keep the pipeline running.
+Currently, the bears CI is broken, so I will start by fixing it.
```
Comment (abhishalya) [inline cEP-0035.md]: Not required.

File: cEP-0035.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+## Objectives
+
+TThe project's primary goal is to enable coala to support newer versions of Python, but this will be attained in multiple smaller steps. The goal is not just to facilitate support for newer versions of Python but also to upgrade the coala to support more recent dependencies. Upgrades will be performed over multiple repositories as the dependencies across the project are closely interlinked. Will perform changes in coala, bears, dependency_mangement, docker, mobans and coala-utils repository. 
+
+## Proposed Approach
+
+### Keeping the CI/CD pipeline green
+
+Throughout the project, it needs to be ensured that the CI/CD pipeline remains unbroken as it is crucial for running the test suite, which is necessary for PR to get merged. This can be ensured by making commits keeping into consideration the pipeline. And make changes over multiple repositories before merging, in case required to keep the pipeline running.
```
Comment (abhishalya) [inline cEP-0035.md]: cEP should only focus on the potential enhancements, and it doesn't have to revolve around the project.

File: cEP-0035.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+Currently, the bears CI is broken, so I will start by fixing it.
+
+### Upgrade dependency_management system 
+
+The dependency_management system provides a Python API for familiar package managers. It provides a generic method to install dependencies in the host operating system. It has been a while since it has been upgraded. Thus it supports only the earlier versions of package managers. Will upgrade the Python, dependencies and versions of the package managers to access newer dependencies using the API. We will enable support for Python3.9 and then move forward, upgrading the dependencies to their latest versions.
```
Comment (abhishalya) [inline cEP-0035.md]: Again same point as above, sentences like `will upgrade, will enable` etc do not seem fit for the cEP. This is basically just a plain list of enhancements, not the project proposal.

File: cEP-0035.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+### Upgrade bears
+
+Currently bears don't support newer pip versions, changes will be made to fix that. This is essential as while installing the latest version of a Python package might fail or install at a slower rate. This will be followed by upgrade in existing dependencies and Python version. Some changes might be required in mobans and coala as well to keep the pipeline up running. The problems associated with bears and their probable solutions are:
```
Comment (abhishalya) [inline cEP-0035.md]: Please improve the wordings as mentioned in previous points, not mentioning them again to reduce clutter. Just try to give factual info here.

**Review by abhishalya — CHANGES_REQUESTED** [review]
This is looking much better than before, just take care of few comments mentioned inline. Also use a spell checker to ensure there are no typos or grammar errors.

**Comment by abhishalya** [comment]
@anshalshukla Please make necessary changes, and rebase. This needs to be merged today.

**Comment by anshalshukla** [comment]
@abhishalya Will do it soon.

**Comment by anshalshukla** [comment]
@abhishalya, I guess this looks fine now?

**Comment by abhishalya** [comment]
Resolve conflicts as well please.

File: cEP-0035.md [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+
+    async def main():
+        print('Hello ...')
+        await asyncio.sleep(1)
+        print('... World!')
+
+    # Python 3.7+
+    asyncio.run(main())
+    ```
+
```
Comment (abhishalya) [inline cEP-0035.md]: Redundant line.

**Review by abhishalya — CHANGES_REQUESTED** [review]

File: cEP-0035.md [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+This cEP describes the upgrades and steps required to enable coala
+to support newer versions of Python. This project has been taken up
+as a [GSoC'21][1] project.
+
+## Introduction
+
+Lack of support for newer versions of Python has been a significant
+hurdle in the development of coala. coala, as of now, supports
+Python up to version 3.6, which is  quite old. Version 3.6 was
```
Comment (abhishalya) [inline cEP-0035.md]:
```suggestion
Python up to version 3.6, which is quite old. Version 3.6 was
```

File: cEP-0035.md [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+
+Circle CI and AppVeyor were used to ensure Windows and MacOS support,
+these were also maintained as a fallback if any of the CI solution gets
+paid alternatively other can be used. But now with Virtual Machines we
+don't have to ensure Windows and Mac Support, also we now have GitHub
+actions which is available free of cost.
+
+### Update Docker Image
+
+Fix docker image to serve latest version of coala and bears.
```
Comment (abhishalya) [inline cEP-0035.md]:
```suggestion
Fix docker image to serve the latest version of coala and bears.
```

**Review by abhishalya — CHANGES_REQUESTED** [review]

**Comment by anshalshukla** [comment]
@abhishalya Done!

**Review by abhishalya — DISMISSED** [review]
Nice work :)

**Comment by abhishalya** [comment]
@anshalshukla You didn't write valid commit short log. Please read the commit guidelines thoroughly.

```
closes coala#191
```

Please take care of these things before writing any commit. This PR is unmergable until everything is done properly.

**Review by bkhanale — APPROVED** [review]

**Comment by anshalshukla** [comment]
@abhishalya Can I get this merged now.

**Review by abhishalya — APPROVED** [review]

**Pull request merged**